### PR TITLE
DA-502 refactor to allow pairing at different levels

### PR DESCRIPTION
**Pull request by Seeker1911**

### CODE REVIEW

File: rdr_client/hpo_assigner.py [hunk 1]
```diff
@@ -98,4 +99,5 @@ def _get_old_hpo(participant):
   arg_parser = get_parser()
   arg_parser.add_argument('file', help='file containing the list of HPOs and participant IDs')
   arg_parser.add_argument('--dry_run', action='store_true')
+  arg_parser.add_argument('--pairing', help='set level of pairing at site')
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: It looks like this is required above. If so, you may want to add required=True here. Alternatively, make it default to awardee.

File: rdr_client/hpo_assigner.py [hunk 2]
```diff
@@ -64,10 +65,10 @@ def main(client):
       if hpo == 'UNSET':
         participant['providerLink'] = []
       else:
-        participant['providerLink'] = [{'primary': True,
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: _ORG_PREFIX is now unused and can be removed?

File: rdr_client/hpo_assigner.py [hunk 3]
```diff
@@ -25,11 +24,13 @@ def main(client):
   num_no_change = 0
   num_updates = 0
   num_errors = 0
+
   with open(client.args.file) as csvfile:
     reader = csv.reader(csvfile)
     for line in reader:
       try:
         participant_id, hpo = [v.strip() for v in line]
+        pairing = client.args.pairing
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: We should have validation that pairing is one of awardee, organization, and site here.

File: rdr_client/hpo_assigner.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 Usage:
   ./run_client.sh --project all-of-us-rdr-prod --account $USER@pmi-ops.org \
-      hpo_assigner.py participant_ids_and_hpos.csv [--dry_run]
+      hpo_assigner.py participant_ids_and_hpos.csv --pairing [site|organization|awardee] [--dry_run]
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: Once this is merged, make sure to update the oncall playbook:

https://docs.google.com/document/d/1FkzFyc_neXn4gAQfdaCoe91mogD4frIAOFtXDeDM-sU/edit#heading=h.jyaxpj74c83k

File: rdr_client/hpo_assigner.py [hunk 5]
```diff
@@ -64,10 +65,10 @@ def main(client):
       if hpo == 'UNSET':
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: Seems like we may wan to delete providerLink, awardee, hpoId, site, and organization before the if here. That way we're only passing up what we want to change?

Also, the _get_old_hpo logic has to change to use pairing... we're now not necessarily changing awardees, and we need to show the change in organization or site if appropriate.

File: rdr_client/hpo_assigner.py (hunk 4 above)
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: This works OK assuming that we are going to get pairing files that look like 

<PARTICIPANT ID>,<AWARDEE/ORGANIZATION/SITE ID>

and that all of the entries have the same level of hierarchy.

That seems OK to me, but we should first check in with Ashley to make sure that ServiceNow is going to start sending us files that look like that.... and that they will actually change to use this.

Comment (danqrodney) [reply rdr_client/hpo_assigner.py]: (Let's hold off on this PR until we hear back about this.)

Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: spoke with the team, this technique is good.

File: rdr_client/hpo_assigner.py (hunk 3 above)
Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: done

File: rdr_client/hpo_assigner.py (hunk 5 above)
Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: not done

File: rdr_client/hpo_assigner.py (hunk 2 above)
Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: yup

File: rdr_client/hpo_assigner.py (hunk 1 above)
Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: done

File: rdr_client/hpo_assigner.py (hunk 4 above)
Comment (danqrodney) [reply rdr_client/hpo_assigner.py]: So is ServiceNow going to start sending us data at org / site level?

File: rdr_client/hpo_assigner.py [hunk 6]
```diff
@@ -13,31 +13,33 @@
...
   num_errors = 0
+
   with open(client.args.file) as csvfile:
     reader = csv.reader(csvfile)
     for line in reader:
       try:
-        participant_id, hpo = [v.strip() for v in line]
+        participant_id, new_pairing = [v.strip() for v in line]
+        pairing = client.args.pairing
+        assert pairing in ('site', 'organization', 'awardee')
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: We should do this check once, up front, and bail out if it fails... not on every line.

File: rdr_client/hpo_assigner.py [hunk 7]
```diff
@@ -13,31 +13,33 @@
...
+      except AssertionError:
+        logging.error('Pairing must be one of site|organization|awardee')
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: Rather than asserting and catching AssertionError, just do an if.

File: rdr_client/hpo_assigner.py (hunk 4 above)
Comment (Seeker1911) [reply rdr_client/hpo_assigner.py]: according to Ashley, yes.

File: rdr_client/hpo_assigner.py [hunk 8]
```diff
@@ -13,31 +15,35 @@
...
+  if client.args.pairing not in pairing_list:
+    sys.exit('Pairing must be one of site|organization|awardee')
+
   with open(client.args.file) as csvfile:
     reader = csv.reader(csvfile)
     for line in reader:
       try:
-        participant_id, hpo = [v.strip() for v in line]
+        participant_id, new_pairing = [v.strip() for v in line]
+        pairing = client.args.pairing
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: You can assign this just once, outside the loop.

Also, rename to "pairing_key"?

File: rdr_client/hpo_assigner.py [hunk 9]
```diff
@@ -1,10 +1,12 @@
-"""Assigns participants with the specified IDs to the test HPO.
+"""Assigns participants with the specified IDs to the organization.
 
 Usage:
   ./run_client.sh --project all-of-us-rdr-prod --account $USER@pmi-ops.org \
-      hpo_assigner.py participant_ids_and_hpos.csv [--dry_run]
+      hpo_assigner.py participant_ids_and_hpos.csv --pairing [site|organization|awardee] [--dry_run]
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: We might want to rename this script "pairing_assigner" or whatever.

File: rdr_client/hpo_assigner.py [hunk 10]
```diff
@@ -54,20 +60,21 @@ def main(client):
...
+        logging.info('%s unchanged (already %s)', participant_id, old_pairing)
         continue
 
-      logging.info('%s %s => %s', participant_id, old_hpo, hpo)
-      if hpo == 'UNSET':
-        participant['providerLink'] = []
+      logging.info('%s %s => %s', participant_id, old_pairing, new_pairing)
+      if new_pairing == 'UNSET':
+        for i in pairing_list:
+          participant[i] = 'UNSET'
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: Is it correct to set site and organization to UNSET? Or do we just null them out?

What about providerLink? Shouldn't we null that out?

File: rdr_client/hpo_assigner.py [hunk 11]
```diff
@@ -54,20 +60,21 @@ def main(client):
...
       else:
-        participant['providerLink'] = [{'primary': True,
-                                        'organization': {'reference': _ORG_PREFIX + hpo}}]
+        participant[pairing] = new_pairing
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: Should we delete providerLink and other pairing list entries?

File: rdr_client/hpo_assigner.py [hunk 12]
```diff
@@ -1,10 +1,12 @@
...
 
 Usage:
   ./run_client.sh --project all-of-us-rdr-prod --account $USER@pmi-ops.org \
-      hpo_assigner.py participant_ids_and_hpos.csv [--dry_run]
+      hpo_assigner.py participant_ids_and_hpos.csv --pairing [site|organization|awardee] [--dry_run]
+Where site = google_group, organization = external_id, awardee = name.
 
-Where the CSV contains lines with P12345678,NEW_HPO_ID like:
+The CSV contains lines with P12345678,NEW_ORGANIZATION like:
+Example awardees:
```
Comment (danqrodney) [inline rdr_client/hpo_assigner.py]: We might want an example for sites, too.

**Comment by Seeker1911** [comment]
Spoke with Ashley, the expected behavior is to clear everything if 'UNSET'.

**Pull request merged**